apps/mobile/src/contexts/GalleryEditor/collectionLayout.ts

```diff
@@ -0,0 +1,196 @@
+};
+
+type Section<T> = {
+  id: string;
+  items: ReadonlyArray<T | WhitespaceBlock>;
+  columns: number;
+};
+type CollectionWithLayout<T> = Array<Section<T>>;
+
+export function parseCollectionLayoutGraphql<T>(
+  tokens: T[],
+  collectionLayoutRef: collectionLayoutParseFragment$key,
+  ignoreWhitespace = false
+): CollectionWithLayout<T> {
+  const layout = readInlineData(
+    graphql`
+      fragment collectionLayoutParseFragment on CollectionLayout @inline {
+        sections
+        sectionLayout {
+          whitespace
+          columns
+        }
+      }
+    `,
+    collectionLayoutRef
+  );
+
+  const sections = removeNullValues(layout?.sections ?? []);
+  const sectionLayout = removeNullValues(layout?.sectionLayout ?? []);
+
+  return parseCollectionLayout(
+    tokens,
+    {
+      sections,
+      sectionLayout,
+    },
+    ignoreWhitespace
+  );
+}
+
+type CollectionLayout = {
+  sections: Array<number>;
+  sectionLayout: Array<{ whitespace: ReadonlyArray<number | null> | null; columns: number | null }>;
+};
+
+// Given a list of tokens in the collection and the collection layout settings,
+// returns an object that represents the full structure of the collection layout with sections, items, and whitespace blocks.
+export function parseCollectionLayout<T>(
+  tokens: ReadonlyArray<T>,
+  collectionLayout: CollectionLayout,
+  ignoreWhitespace = false
+): CollectionWithLayout<T> {
+  if (tokens.length === 0) {
+    return [{ id: generate12DigitId(), items: [], columns: DEFAULT_COLUMNS }];
+  }
+
+  const parsedCollection = collectionLayout.sections.reduce(
+    (allSections: CollectionWithLayout<T>, sectionStartIndex: number, index: number) => {
+      const nextSection = collectionLayout.sections[index + 1];
+      const sectionEndIndex = nextSection ? nextSection - 1 : tokens.length;
+
+      let section: ReadonlyArray<T | WhitespaceBlock> = tokens.slice(
+        sectionStartIndex,
+        sectionEndIndex + 1
+      );
+      if (!ignoreWhitespace) {
+        section = insertWhitespaceBlocks(
+          section,
+          removeNullValues(collectionLayout.sectionLayout[index]?.whitespace)
+        );
+      }
+      const sectionId = generate12DigitId();
+
+      allSections.push({
+        id: sectionId,
+        items: section,
+        columns: collectionLayout.sectionLayout[index]?.columns ?? 1,
+      });
+
+      return allSections;
+    },
+    []
+  );
+
+  return parsedCollection;
+}
+
+// Each value in the whitespace list represents the index of the NFT that a whitespace appears before.
+// For example, whitespace: [0] means that there is one whitespace before the first NFT in the collection.
+// Here is a more detailed example:
+// If Staged items looks like this: (x is whitespace) [x, x, A, x, B, C, D, x, E, x]
+// The list of NFTs would be: [A, B, C, D, E]
+// The whitespace list would be : [0, 0, 1, 4, 5]
+
+export function getWhitespacePositionsFromStagedItems(stagedItems: StagedItem[]): number[] {
```

Comment:
could these functions live in /shared?

Reply:
Yes we could! Might need to update it since `web` and `mobile` use different types. 
https://github.com/gallery-so/gallery/blob/jakz%2Fgallery-editor/apps/mobile/src/contexts/GalleryEditor/types.ts

Reply:
Moved most of it to shared!

Reply:
oh good call, we'd have to use typescript generics to make the types compatible. can handle later